airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -385,3 +404,44 @@ def state(self) -> MutableMapping[str, Any]:
     def state(self, value: StreamState):
         """State setter, accept state serialized by state getter."""
         self.stream_slicer.update_cursor(value)
+
+    def parse_records_and_emit_request_and_responses(self, request, response, stream_slice, stream_state) -> Iterable[AirbyteMessage]:
+        yield self._create_trace_message_from_request(request)
+        yield self._create_trace_message_from_response(response)
+        # Not great to need to call _read_pages which is a private method
+        # A better approach would be to extract the HTTP client from the HttpStream and call it directly from the HttpRequester
+        for record_mapping in self._read_pages(
+            lambda req, res, state, _slice: self.parse_response(res, stream_slice=_slice, stream_state=state), stream_slice, stream_state
+        ):
+            # FIXME: need to get the real json schema if we want to support type transforms
+            json_schema = {}
+            yield stream_data_to_airbyte_message(self.name, record_mapping, self.transformer, json_schema)
+
+    def _create_trace_message_from_request(self, request: requests.PreparedRequest):
+        # FIXME: this should return some sort of trace message
+        request_object = Request(url=request.url, headers=request.headers, body=self._safe_request_body(request.body))
+        log_message = filter_secrets(f"request:{str(request_object)}")
```

Comment:
It would be easier for the `connector-builder-server` to interpret the log message and transform the string back into its respective fields (url, body, headers) if it was formatted as a JSON string. From the connector server we can just perform a `json.loads()`. We could still have the Request/Response dataclasses, but might not be necessary since we'll be emitting it as JSON instead anyway

From the CDK side:
```
request_dict = {
  "url": "https://harvest.greenhouse.io/v1/candidates?per_page=100",
  "headers": { "User-Agent": "python-requests/2.28.1" },
  "body": {}
}
filter_secrets(f"request:{ json.dumps(request_dict)}")
```

And from the connector builder side we would unpack it using some form of:
```
raw_request = log_message.message.partition("request:")[2]
try:
  return json.loads(raw_request)
```

Reply:
good call. I updated the format.

I used the following code to test the parsing:
```
        request_dict = {"url": request.url, "headers": dict(request.headers), "body": self._safe_request_body(request.body)}
        log_message = filter_secrets(f"request:{json.dumps(request_dict)}")
        raw_request = log_message.partition("request:")[2]
        print(json.loads(raw_request))
        exit()
```

and the output was:
```
{'url': 'https://api.courier.com/messages?page_size=1', 'headers': {'User-Agent': 'python-requests/2.28.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Bearer ****'}, 'body': {}}
```